### ISSUE TITLE
No text on UI elements in big picture mode

### ISSUE THREAD

**Issue by DefaultX-od** [canonical/steam-snap#5]
![no ui text](https://user-images.githubusercontent.com/61311905/166107466-3e4b6db1-f66a-40ef-8a97-39acbcda2e74.png)

**Comment by Nightmayr** [comment]
Can confirm I see this as well. When running from the command line I get the following output:

```
(process:10247): Pango-WARNING **: failed to choose a font, expect ugly output. engine-type='PangoRenderFc', script='latin'

(process:10247): Pango-WARNING **: failed to choose a font, expect ugly output. engine-type='PangoRenderFc', script='common'

(process:10247): Pango-CRITICAL **: pango_ft2_render_transformed: assertion 'PANGO_FT2_IS_FONT (font)' failed

(process:10247): Pango-CRITICAL **: pango_ft2_render_transformed: assertion 'PANGO_FT2_IS_FONT (font)' failed

(process:10247): Pango-CRITICAL **: pango_ft2_render_transformed: assertion 'PANGO_FT2_IS_FONT (font)' failed

(process:10247): Pango-CRITICAL **: pango_ft2_render_transformed: assertion 'PANGO_FT2_IS_FONT (font)' failed
```

The critical errors keep repeating as you move around the UI.

**Comment by mhalano** [comment]
I observed this on the menu where I configure the gamepad

**Comment by whitty** [comment]
Same problem on 20.04 - tried `--beta` first, then `--edge` - both showed the problem

```
tracking:     latest/edge
refresh-date: today at 21:26 AEST
channels:
  latest/stable:    –                              
  latest/candidate: –                              
  latest/beta:      1.0.0.74 2022-05-17 (17) 269MB -
  latest/edge:      1.0.0.74 2022-06-10 (27) 265MB -
installed:          1.0.0.74            (27) 265MB -
```

```
$ snap version
snap    2.56
snapd   2.56
series  16
ubuntu  20.04
kernel  5.11.0-44-generictermter
```

**Comment by mhalano** [comment]
The problem still occurs in the latest versions.
```
tracking:     latest/edge
refresh-date: today at 03:28 -03
channels:
  latest/stable:    –                              
  latest/candidate: –                              
  latest/beta:      1.0.0.74 2022-05-17 (17) 269MB -
  latest/edge:      1.0.0.74 2022-07-15 (32) 283MB -
```
```
$ snap version   
snap    2.56.3+git639.g2d00acf
snapd   2.56.3+git639.g2d00acf
series  16
ubuntu  22.10
kernel  5.19.0-9-generic
```

**Comment by ashuntu** [comment]
Not sure why this issue was closed, reopening.

**Comment by whitty** [comment]
Minimal googling of `PangoRenderFc` and the error messages suggests this could be related to fontconfig.  I did notice there is a separate `etc/fonts/fonts.conf` specifically for big-picture (I'm assuming its also called "panorama") `.local/share/Steam/bin/panorama/etc/fonts`

I played around a little, but I realise I don't know anything about how to drive fontconfig.  Does anyone know how to merge fontconfig - er - configs?

Or just try to prove the point by replacing the global snap fontconfig with just the panorama one (maybe main app fonts will break and big-picture work - at least proving the cause)

Basically I'm out of my pay-grade - does anyone have other ideas?

**Comment by ZoopOTheGoop** [comment]
You can't clobber the fonts in the Steam directories, turns out. Steam will immediately rewrite them on launch, sadly. So ~some other debugging method~ is needed.

**Comment by ZoopOTheGoop** [comment]
After doing a little bit of `FC_DEBUG `tracing on both the deb and snap version of Steam, I found out Steam *does* distribute the necessary fonts, it's under `Steam/tenfoot/resource/fonts`, there's two files that are plainly labeled as `DroidSansThai-Regular.ttf` and `NotoSansCJKjp-Regular.otf`, but if you run `FC_DEBUG` to trace loading the pattern matches in the debian version, you'll realize that it's also loading `Motiva Sans` as `$HOME/.steam/debian-installation//tenfoot/resource/fonts/tenfoot.uifont:8`

This file is also in the snap under roughly the same place (`$SNAP_USER_COMMON/.local/share/Steam/tenfoot/...`) and a `diff` exits clean, so the file is correct

So everything exits cleanly, the question is why the Snap version isn't picking up this directory.

This is very hard to answer because debug traces point to a Linux UI rendering file in `panorama` which is the UI rendering library in Source 2. This is basically just like QT or GTK or whatever, you can actually play with it (closed source - as in make your own UIs as client code) with DOTA 2 custom map editor and it lets you use some custom js and css for markup from what I understand. 

This is all in the `tenfoot` directory (with a cousin for the main client in the `steamui` sister directory), and the answer may be there if I do a more thorough audit, but just searching for any font-related stuff it's all fairly generic path-relative markup - i.e. it's not like they hardcoded an (obvious) absolute path anywhere. It's expecting the font to be picked up naturally meaning there's likely somewhere else (whether that be in the proprietary `panorama` code, or in the Steam environment setup) where it expects `tenfoot/resources/fonts` to be appended to the `fontconf` path or the fonts staged in the `fontconf` cache or similar and it's just not.

I think this gives us a much stronger lead on the actual issue, but no actual answers yet. I think this is probably a configuration issue on our end, but there is a small chance this comes down to some path handling that doesn't account for the slight differences in where the Steam root is on a native or flatpak install vs a snap install.

**Comment by ZoopOTheGoop** [comment]
Found it, in `steamui/css/7936.css`

```css
@font-face{font-family:"Motiva Sans";src:url("/custom_fonts/clientui.uifont?MotivaSans-Regular") format("truetype")}@
```

(Among a bunch of other stuff, this is repeated for other fonts). Expanding the search to the entire Steam root, this shows up in `friends.css` and `fonts.css` in `clientui` (which appears to be the friends list). I haven't found anything else that mentions `/custom_fonts` (nor in the steam runtime (-tools)) which leads me to believe that perhaps `panorama` has a directory structure it expects and then exposes to its client code in a specific way. My lack of knowledge about web development probably hurts me here.

My best guess is at some point `XDG_DATA_HOME` is set to `{tenfoot|clientui|steamui}/resources` before launching the matching service to load it, since that would line up with the `fonts.conf` in `panorama` having `<dir prefix="xdg">fonts</dir>` in it, and all the `resources` have a `/fonts` directory. This would mean that for ~some reason~ in the Snap, the transition between the regular client and Big Picture Mode isn't properly changing `XDG_DATA_HOME`. 

The counterpoint to this hypothesis is this would that the `clientui` (i.e. Friends list`) also specifies `Motiva Sans`, but the counter-counterpoint is that every *actual* instance in both `steamui` and `clientui` `Motiva <X>` is specified, a fallback to Arial, Helvetica, and finally generic `sans-serif` is provided, while in `Big Picture` it is not. This would imply that even were `XDG_DATA_HOME` set improperly, the first two would fall back to proper fonts, while Big Picture, specifically, wouldn't, which is consistent with the behavior we're seeing. However, the counter-counterpoint is that these UIs all successfully download images to their resources directory, so if it exclusively relied on` XDG_DATA_HOME` that would fail too - but it's possible it's smart enough to handle that but `fontconf` itself isn't.

Searching around, it seems like people have had issues with `$XDG_DATA_HOME` in snap before, but nothing similar to this. By default it's set to `$HOME/snap/steam/common/.local/share` in the Steam snap shell, and manually setting it doesn't immediately cause `snapd` to change it back or anything obvious like that, but I think this is probably the ballpark of the issue. That or I've learned an impressive amount about `fontconf` to chase a very big red herring. One of the other.

**Comment by PlatFormPlayZ** [comment]
> Found it, in `steamui/css/7936.css`
> 
> ```css
> @font-face{font-family:"Motiva Sans";src:url("/custom_fonts/clientui.uifont?MotivaSans-Regular") format("truetype")}@
> ```
> 
> (Among a bunch of other stuff, this is repeated for other fonts). Expanding the search to the entire Steam root, this shows up in `friends.css` and `fonts.css` in `clientui` (which appears to be the friends list). I haven't found anything else that mentions `/custom_fonts` (nor in the steam runtime (-tools)) which leads me to believe that perhaps `panorama` has a directory structure it expects and then exposes to its client code in a specific way. My lack of knowledge about web development (and webkit specifically) probably hurts me here.
> 
> My best guess is at some point `XDG_DATA_HOME` is set to `{tenfoot|clientui|steamui}/resources` before launching the matching service to load it (probably in a runtime or pressure vessel), since that would line up with the `fonts.conf` in `panorama` having `<dir prefix="xdg">fonts</dir>` in it, and all the `resources` have a `/fonts` directory. This would mean that for ~some reason~ in the Snap, the transition between the regular client and Big Picture Mode isn't properly changing `XDG_DATA_HOME`. I do remember seeing in the runtime docs that at least some of them set some similar `XDG` values before wrapping so that games know where they're supposed to write to/live (or at least experimented with that at some point).
> 
> The counterpoint to this hypothesis is this would that the `clientui` (i.e. Friends list) also specifies `Motiva Sans`, but the counter-counterpoint is that every _actual_ instance in both `steamui` and `clientui` `Motiva <X>` is specified, a fallback to Arial, Helvetica, and finally generic `sans-serif` is provided, while in `Big Picture` it is not. This would imply that even were `XDG_DATA_HOME` set improperly, the first two would fall back to proper fonts, while Big Picture, specifically, wouldn't, which is consistent with the behavior we're seeing. However, the counter-counterpoint is that these UIs all successfully download images to their resources directory, so if it exclusively relied on` XDG_DATA_HOME` that would fail too - but it's possible it's smart enough to handle that but `fontconfig` itself isn't.
> 
> Searching around, it seems like people have had issues with `$XDG_DATA_HOME` in snap before, but nothing similar to this. By default it's set to `$HOME/snap/steam/common/.local/share` in the Steam snap shell, and manually setting it doesn't immediately cause `snapd` to change it back or anything obvious like that, but I think this is probably the ballpark of the issue. That or I've learned an impressive amount about `fontconfig` and panorama to chase a very big red herring. One or the other.
> 
> One other note is that this seems to be a normal webkit way to load fonts from a source `uifont` file, but the whole `/custom_fonts` mapping remains elusive hence the `XDG_DATA_HOME` musing. It implies some sort of artificial environment is being created, since no such directory can be found and basic webkit/electron tutorials don't mention anything similar (they just use relative paths or a non-obfuscated project root typically).

knowing this, as someone who is relatively new to linux, do you know of a way to fix said issue? (even if it is a client side fix)

**Comment by whitty** [comment]
> knowing this, as someone who is relatively new to linux, do you know of a way to fix said issue? (even if it is a client side fix)

I don't know how to fix it - but to investigate you can:

1. Start steam (snap)
2. get a shell up in that snap `snap run --shell steam`
3. Find the running steam executable's pid `ps fax` or `top` might help
4. Dump the environment for that pid:   `cat  /proc/<pid>/environ  | tr "\0" "\n"`  (where you substitute the pid found in step 3 for `<pid>`)
5. Switch to big picture mode - look to see if there's a new pid (probably)
6. dump that environment and compare
7. repeat steps 3-6 above in non-snapped install and see if there's any obvious difference - or if something is pointing to an odd non-existent directory

**Comment by ZoopOTheGoop** [comment]
Okay, it's the `desktop-launch` script (taken from the `gnome` snapcraft extension). It's written in a way that breaks fontconfig for at least some application (it breaks `wine` too, for instance).

If you want a workaround for now to run steam run:

```bash
snap run --shell steam
unset C
unset FONTCONFIG_FILE
steam
```

And it will work (not: this may break localization for non-English versions of Steam, I'm not sure).
